src/Factory.sol

```diff
@@ -0,0 +1,172 @@
+        /// @dev ensure one account per address
+        if (ownerToAccount[msg.sender] != address(0)) {
+            revert OnlyOneAccountPerAddress(ownerToAccount[msg.sender]);
+        }
+
+        // create account and set beacon to this address (i.e. factory address)
+        accountAddress = payable(address(new AccountProxy(address(this))));
+
+        // update owner to account mapping
+        ownerToAccount[msg.sender] = accountAddress;
+
+        // initialize new account
+        (bool success, bytes memory data) = accountAddress.call(
+            abi.encodeWithSignature(
+                "initialize(address,address,address,address,address,address)",
+                msg.sender, // caller will be set as owner
+                marginAsset,
+                addressResolver,
+                settings,
+                ops,
+                address(this)
+            )
+        );
+        if (!success) revert AccountFailedToInitialize(data);
+
+        // determine version for the following event
+        (success, data) = accountAddress.call(
+            abi.encodeWithSignature("VERSION()")
+        );
+        if (!success) revert AccountFailedToFetchVersion(data);
+
+        emit NewAccount({
+            creator: msg.sender,
+            account: accountAddress,
+            version: abi.decode(data, (bytes32))
+        });
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                           ACCOUNT OWNERSHIP
+    //////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IFactory
+    function updateAccountOwner(address _oldOwner, address _newOwner)
+        external
+        override
+    {
+        /// @dev ensure _newOwner does not already have an account
+        if (ownerToAccount[_newOwner] != address(0)) {
+            revert OnlyOneAccountPerAddress(ownerToAccount[_newOwner]);
+        }
+
+        // get account address
+        address account = ownerToAccount[_oldOwner];
+
+        // ensure account exists
+        if (account == address(0)) revert AccountDoesNotExist();
+
+        // ensure account owned by _oldOwner is the caller
+        if (msg.sender != account) revert CallerMustBeAccount();
+
+        delete ownerToAccount[_oldOwner];
+        ownerToAccount[_newOwner] = account;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                             UPGRADABILITY
+    //////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IFactory
+    function upgradeSystem(
+        address _implementation,
+        address _settings,
+        address _marginAsset,
+        address _addressResolver,
+        address payable _ops
+    ) external override onlyOwner {
+        if (!canUpgrade) revert CannotUpgrade();
+
+        implementation = _implementation;
+        settings = _settings;
+        marginAsset = _marginAsset;
+        addressResolver = _addressResolver;
+        ops = _ops;
```

Comment:
Would these only affect new accounts? If so, this prob isn't the best idea to bundle with implementation upgrade.

Implementation upgrade -> all accounts
Settings, margin asset, address resolver, ops -> future accounts? 

best to have separate function for this or to put these things on the account impl itself

Reply:
Good idea. Will separate this out

Reply:
🏁🏁🏁